### ISSUE TITLE
Got a promise instead of imported object (after the last alpha update)

### ISSUE THREAD

**Issue by smelukov** [module-federation/vite#47]
@zhangHongEn

I have a code like this:

```ts
import {Some} from '@foo/bar';

Some.create();
```

`@foo/bar` is in `shared` section with `singleton: true`

After the last alpha update I've got a Promise in `Some`.
Transpiled code is something like:

```js
var bar_1 = require_my_app_mf_loadShare_foo_bar();

// ...

bar_1.Some.create();
```

So, `bar_1` is a promise

`require_my_app_mf_loadShare_foo_bar` is a function that was generated by this code https://github.com/module-federation/vite/blob/new-config/src/virtualModules/virtualShared_preBuild.ts#L56

**Comment by zhangHongEn** [comment]
Do all modules shared by this project get promise?

**Comment by zhangHongEn** [comment]
I need @foo_bar.js in the network, for example react-dom.js in my network
``` js
import {require_index_cjs} from "/node_modules/.vite/deps/chunk-WVZCKULC.js?v=6fa77d23";
import {__commonJS} from "/node_modules/.vite/deps/chunk-BUSYA2B4.js?v=6fa77d23";

// ../../../node_modules/.pnpm/an-empty-js-file@1.0.0/node_modules/an-empty-js-file/viteViteHost___mf__loadShare_react-dom.js
var require_viteViteHost_mf_loadShare_react_dom = __commonJS({
    "../../../node_modules/.pnpm/an-empty-js-file@1.0.0/node_modules/an-empty-js-file/viteViteHost___mf__loadShare_react-dom.js"(exports, module) {
        var {loadShare} = require_index_cjs();
        var res = loadShare("react-dom", {
            customShareInfo: {
                shareConfig: {
                    singleton: false,
                    strictVersion: false,
                    requiredVersion: "^18.3.1"
                }
            }
        });
        var exportModule = (/*mf top-level-await placeholder replacement mf*/
        res.then((factory)=>factory()));
        module.exports = exportModule;
    }
});

const __mfproxy__awaitdefault = await (require_viteViteHost_mf_loadShare_react_dom());
const __mfproxy__default = __mfproxy__awaitdefault;

export {__mfproxy__default as default};
```

**Comment by husayt** [comment]
@zhangHongEn I get this as well with Material UI:

The following code from MUI ( node_modules/@mui/styled-engine/index.js)
```
import emStyled from '@emotion/styled';
export default function styled(tag, options) {
  const stylesFactory = emStyled(tag, options);
  if (process.env.NODE_ENV !== 'production') {
    return (...styles) => {
      const component = typeof tag === 'string' ? `"${tag}"` : 'component';
      if (styles.length === 0) {
        console.error([`MUI: Seems like you called \`styled(${component})()\` without a \`style\` argument.`, 'You must provide a `styles` argument: `styled("div")(styleYouForgotToPass)`.'].join('\n'));
      } else if (styles.some(style => style === undefined)) {
        console.error(`MUI: the styled(${component})(...args) API requires all its args to be defined.`);
      }
      return stylesFactory(...styles);
    };
  }
  return stylesFactory;
}
```
becomes like this
```
function styled(tag, options) {
  const stylesFactory = (0, import_styled.default)(tag, options);
  if (true) {
    return (...styles) => {
      const component = typeof tag === "string" ? `"${tag}"` : "component";
      if (styles.length === 0) {
        console.error([`MUI: Seems like you called \`styled(${component})()\` without a \`style\` argument.`, 'You must provide a `styles` argument: `styled("div")(styleYouForgotToPass)`.'].join("\n"));
      } else if (styles.some((style4) => style4 === void 0)) {
        console.error(`MUI: the styled(${component})(...args) API requires all its args to be defined.`);
      }
      return stylesFactory(...styles);
    };
  }
  return stylesFactory;
}

where `import_styled.default` is a promise

**Comment by zhangHongEn** [comment]
/*mf top-level-await placeholder replacement mf*/

I need to check if this line of comment exists in the code in the network request

---- Replied Message ----
| From | ***@***.***> |
| Date | 08/12/2024 01:01 |
| To | module-federation/vite ***@***.***> |
| Cc | zhn ***@***.***>,
Mention ***@***.***> |
| Subject | Re: [module-federation/vite] Got a promise instead of imported object (after the last alpha update) (Issue #47) |

@zhangHongEn I get this as well with Material UI:

The following code from MUI ( ***@***.***/styled-engine/index.js)

import emStyled from ***@***.***/styled';
export default function styled(tag, options) {
  const stylesFactory = emStyled(tag, options);
  if (process.env.NODE_ENV !== 'production') {
    return (...styles) => {
      const component = typeof tag === 'string' ? `"${tag}"` : 'component';
      if (styles.length === 0) {
        console.error([`MUI: Seems like you called \`styled(${component})()\` without a \`style\` argument.`, 'You must provide a `styles` argument: `styled("div")(styleYouForgotToPass)`.'].join('\n'));
      } else if (styles.some(style => style === undefined)) {
        console.error(`MUI: the styled(${component})(...args) API requires all its args to be defined.`);
      }
      return stylesFactory(...styles);
    };
  }
  return stylesFactory;
}


becomes like this

function styled(tag, options) {
  const stylesFactory = (0, import_styled.default)(tag, options);
  if (true) {
    return (...styles) => {
      const component = typeof tag === "string" ? `"${tag}"` : "component";
      if (styles.length === 0) {
        console.error([`MUI: Seems like you called \`styled(${component})()\` without a \`style\` argument.`, 'You must provide a `styles` argument: `styled("div")(styleYouForgotToPass)`.'].join("\n"));
      } else if (styles.some((style4) => style4 === void 0)) {
        console.error(`MUI: the styled(${component})(...args) API requires all its args to be defined.`);
      }
      return stylesFactory(...styles);
    };
  }
  return stylesFactory;
}

where `import_styled.default` is a promise

**Comment by zhangHongEn** [comment]
I'll try @ mui later

---- Replied Message ----
| From | ***@***.***> |
| Date | 08/12/2024 01:01 |
| To | module-federation/vite ***@***.***> |
| Cc | zhn ***@***.***>,
Mention ***@***.***> |
| Subject | Re: [module-federation/vite] Got a promise instead of imported object (after the last alpha update) (Issue #47) |

@zhangHongEn I get this as well with Material UI:

The following code from MUI ( ***@***.***/styled-engine/index.js)

import emStyled from ***@***.***/styled';
export default function styled(tag, options) {
  const stylesFactory = emStyled(tag, options);
  if (process.env.NODE_ENV !== 'production') {
    return (...styles) => {
      const component = typeof tag === 'string' ? `"${tag}"` : 'component';
      if (styles.length === 0) {
        console.error([`MUI: Seems like you called \`styled(${component})()\` without a \`style\` argument.`, 'You must provide a `styles` argument: `styled("div")(styleYouForgotToPass)`.'].join('\n'));
      } else if (styles.some(style => style === undefined)) {
        console.error(`MUI: the styled(${component})(...args) API requires all its args to be defined.`);
      }
      return stylesFactory(...styles);
    };
  }
  return stylesFactory;
}


becomes like this

function styled(tag, options) {
  const stylesFactory = (0, import_styled.default)(tag, options);
  if (true) {
    return (...styles) => {
      const component = typeof tag === "string" ? `"${tag}"` : "component";
      if (styles.length === 0) {
        console.error([`MUI: Seems like you called \`styled(${component})()\` without a \`style\` argument.`, 'You must provide a `styles` argument: `styled("div")(styleYouForgotToPass)`.'].join("\n"));
      } else if (styles.some((style4) => style4 === void 0)) {
        console.error(`MUI: the styled(${component})(...args) API requires all its args to be defined.`);
      }
      return stylesFactory(...styles);
    };
  }
  return stylesFactory;
}

where `import_styled.default` is a promise

**Comment by smelukov** [comment]
@zhangHongEn 

I have this structure:
`@foo/bar` - shared dep
`@some/library` - npm library that uses `@foo/bar`
`my-app` - host that uses `@some/library`

`@some/library` chunk has this code (and a lot of business logic):

```
// ../../node_modules/an-empty-js-file/my__app___mf__loadShare_foo__bar.js
var require_my_app_mf_loadShare_foo_bar = __commonJS({
    "../../node_modules/an-empty-js-file/my__app___mf__loadShare_foo__bar.js"(exports, module) {
        var {loadShare} = require_index_cjs();
        var res = loadShare("@foo/bar", {
            customShareInfo: {
                shareConfig: {
                    singleton: true,
                    strictVersion: false,
                    requiredVersion: "^0.22.1"
                }
            }
        });
        var exportModule = (/*mf top-level-await placeholder replacement mf*/
        res.then((factory)=>factory()));
        module.exports = exportModule;
    }
});
```

`@foo/bar` chunks has a lot of business logic

**Comment by smelukov** [comment]
@zhangHongEn 
https://localhost:9001/node_modules/.vite/deps/__mf__prebuildwrap_foo__bar__js.js?v=5028c11a

Does not have any of these

**Comment by smelukov** [comment]
`https://localhost:9001/node_modules/.vite/deps/@some_library.js?v=aa0f7309` has this code:
```js
// ../../node_modules/an-empty-js-file/my__app___mf__loadShare_foo__bar.js
var require_my_app_mf_loadShare_foo_bar = __commonJS({
    "../../node_modules/an-empty-js-file/my__app___mf__loadShare_foo__bar.js"(exports, module) {
        var {loadShare} = require_index_cjs();
        var res = loadShare("@foo/bar", {
            customShareInfo: {
                shareConfig: {
                    singleton: true,
                    strictVersion: false,
                    requiredVersion: "^0.22.1"
                }
            }
        });
        var exportModule = (/*mf top-level-await placeholder replacement mf*/
        res.then((factory)=>factory()));
        module.exports = exportModule;
    }
});
```

**Comment by zhangHongEn** [comment]
> @zhangHongEn `https://localhost:9001/node_modules/.vite/deps/__mf__prebuildwrap_foo__bar__js.js?v=5028c11a`
> 
> Does not have any of these

![image](https://github.com/user-attachments/assets/ca13505c-7503-4b29-9c46-edce5f126b31)

You are right, it has become chunk-hash.js
Can you use search to find its code in the network? I have not reproduced this problem yet, and need to find it to troubleshoot
``` js
loadShare("@foo/bar"
```

**Comment by smelukov** [comment]
@zhangHongEn the only place with `loadShare("@foo/bar"` is https://github.com/module-federation/vite/issues/47#issuecomment-2283509622

**Comment by husayt** [comment]
@smelukov is it possible to get this reproduced with examples we have and then we can get to resolve that.

I have done the same for image loading problem from remote, see:

![image](https://github.com/user-attachments/assets/a680d321-5dd9-4ed8-8c73-b4a056319129)

**Comment by zhangHongEn** [comment]
> @zhangHongEn the only place with `loadShare("@foo/bar"` is [#47 (comment)](https://github.com/module-federation/vite/issues/47#issuecomment-2283509622)

It lacks the export code of esmodule. In theory, such a file should not exist.

@husayt  mentioned that @mui seems to have the same problem. I will try it using @mui tonight.

**Comment by smelukov** [comment]
@zhangHongEn @husayt I've made a reproduction example https://github.com/smelukov/vite/tree/repro (repro-branch)

exec `npm run dev` in `examples/vite-vite/vite-host` and you'll see:

```
Uncaught TypeError: events_1.Emitter is not a constructor
    at Initializer.create (@some_library.js?v=8f723edf:35:21)
    at main.jsx:7:25
```

**Comment by zhangHongEn** [comment]
> @zhangHongEn @husayt I've made a reproduction example https://github.com/smelukov/vite/tree/repro (repro-branch)
> 
> * exec `npm run build` in `examples/vite-vite/some`
> * exec `npm run dev` in `examples/vite-vite/vite-host` and you'll see:
> 
> ```
> Uncaught TypeError: events_1.Emitter is not a constructor
>     at Initializer.create (@some_library.js?v=8f723edf:35:21)
>     at main.jsx:7:25
> ```
> 
> Maybe it's because some/library is not esm library?

Not sure why, react is also a cjs library, I will check it tonight

**Comment by smelukov** [comment]
@zhangHongEn can we use https://github.com/zhangHongEn/vite/commit/629799180835669498c6768236525af6c45a756c ?

**Comment by zhangHongEn** [comment]
> @zhangHongEn can we use [zhangHongEn@6297991](https://github.com/zhangHongEn/vite/commit/629799180835669498c6768236525af6c45a756c) ?

Yes, I found another problem. There is a probability of failure to obtain the shared module in dev mode. Refreshing the page again works fine and does not affect the build. I originally wanted to fix it together and submit a PR.

**Comment by smelukov** [comment]
Is there a published version to use the fix?

**Comment by smelukov** [comment]
And thanks for maintaining the plugin!

**Comment by zhangHongEn** [comment]
#53 
Waiting for this to be merged, thank you for finding these problems

**Comment by gioboa** [comment]
Closed by #53
